3096 fix config listener accumulation

davisp opened the pull request:
This fixes an issue we've found with event_handlers not being removed from config_event because they're registered in the config gen_server which never exits. This fixes the issue by creating a monitor process that maintains the current callback API that we have rather than reverting to the half callback/half message passing callback.

davisp commented:
I should mention that the end result of this bug is that handlers are never removed. On a busy cluster we were seeing tens of millions of handlers which led to config_event using ten to fifteen gigabytes of RAM which is mmmmbad.

kxepal commented:
Wow! Nice found!

iilyak commented on src/config_listener_mon.erl:
```diff
@@ -0,0 +1,84 @@
+
+init({Pid, Mod, InitSt}) ->
+    Ref = erlang:monitor(process, Pid),
+    case config_listener:start(Mod, {Mod, Pid}, {Pid, InitSt}) of
```
Currently we pass InitState to Module:handle_config_change. With this state we would start passing `{Pid, InitSt}` which has two problems:
- it breaks current contract
- it leaks internals

davisp replied:
I agree that it leaks internals but it doesn't change anything on the current behavior. I had to do this to match the old behavior where we were passing {Subscriber, InitState} with the gen_event handler so that the Subscriber pid could be passed to handle_config_terminate (weirdly). I just moved it along with the call to config_listener:start/3 rather than keeping it wrapped in config.erl's call to config_listener_mon:start_link/2.

iilyak replied:
I was wrong it wouldn't break the contract. Since we do pass Pid already 

[Reply = config_listener:start(CallbackModule, {Subscriber, InitialState})](https://github.com/apache/couchdb-config/blob/master/src/config.erl#L264)

The config_listener would make sure Pid is not passed to [Module:handle_config_change](https://github.com/apache/couchdb-config/blob/master/src/config_listener.erl#L43:L44)

```
handle_event({config_change, Sec, Key, Value, Persist}, {Module, {From, State}}) ->
                                                                   |       |
                                                                   x       |
                                                                           V
             case Module:handle_config_change(Sec, Key, Value, Persist, State) of
```

iilyak commented:
@davisp: Is there any reason why we cannot rely on gen_event provided monitoring. The following change passes the test suite:
```
--- a/src/config.erl
+++ b/src/config.erl
@@ -183,7 +183,7 @@ delete(Section, Key, Persist, Reason) when is_list(Section), is_list(Key) ->


 listen_for_changes(CallbackModule, InitialState) ->
-    config_listener_mon:start_link(CallbackModule, InitialState).
+    config_listener:start(CallbackModule, InitialState).
```

```
--- a/src/config_listener.erl
+++ b/src/config_listener.erl
@@ -35,7 +35,9 @@ start(Module, State) ->
     start(Module, Module, State).

 start(Module, Id, State) ->
-    gen_event:add_sup_handler(config_event, {?MODULE, Id}, {Module, State}).
+    Pid = self(),
+    gen_event:add_sup_handler(config_event,
+        {?MODULE, {Id, Pid}}, {Module, {Pid, State}}).
```

```
--- a/test/config_tests.erl
+++ b/test/config_tests.erl
@@ -461,10 +461,6 @@ should_remove_handler_when_pid_exits(Pid) ->
     ?_test(begin
         ?assertEqual(2, n_handlers()),

-        % Monitor the config_listener_mon process
-        {monitored_by, [Mon]} = process_info(Pid, monitored_by),
-        MonRef = erlang:monitor(process, Mon),
-
         % Kill the process synchronously
         PidRef = erlang:monitor(process, Pid),
         exit(Pid, kill),
@@ -474,14 +470,6 @@ should_remove_handler_when_pid_exits(Pid) ->
             erlang:error({timeout, config_listener_death})
         end,

-        % Wait for the config_listener_mon process to
-        % exit to indicate the handler has been removed.
-        receive
-            {'DOWN', MonRef, _, _, _} -> ok
-        after ?TIMEOUT ->
-            erlang:error({timeout, config_listener_mon_death})
-        end,
-
         ?assertEqual(1, n_handlers())
     end).
```

iilyak commented:
`gen_event` establishes [a link](https://github.com/erlang/otp/blob/maint/lib/stdlib/src/gen_event.erl#L444) and [removes the handler](https://github.com/erlang/otp/blob/maint/lib/stdlib/src/gen_event.erl#L263) in [handle_exit](https://github.com/erlang/otp/blob/maint/lib/stdlib/src/gen_event.erl#L335:L353).

davisp commented:
But wouldn't that completely undercut the entire change to avoid having config_event send a message to the process that calls `config:listen_for_changes/2`?

iilyak commented:
I see. We want to isolate the caller from config_event. So the config_event's exit for example do not propagate to the process that calls config:listen_for_changes. So want to re-implement monitoring based on monitor rather on link.

davisp commented:
@iilyak I'm not sure what you mean by that last sentence. The previous change that introduced this issue was focused on removing the split between message passing and module callbacks. To do that we need a separate process. In the first case that was the config gen_event which turned out to be bad because it never dies so handlers are never removed. This just creates a new config_listener_mon that exits when the requesting process does as a way to trigger the handler removal.

davisp commented:
Oh, I guess config_listener_mon also uses a monitor back to the requesting process rather than a link. I did that because its slightly cleaner than trapping exits in the monitor as well as it doesn't trigger a message anywhere if it exits.

iilyak commented on src/config_listener_mon.erl:
```diff
@@ -0,0 +1,84 @@
+}).
+
+
+start_link(Module, InitSt) ->
```
A nitpick. This a little confusing. The function is called `start_link`. `start_link` usually returns `{ok, Pid}`. In this case we return `ok`. We could either:
- make it return `{ok, Pid}`
- rename it to something else (`subscribe` | `monitor` | `listen`)

iilyak commented:
@davisp: Other than a [little nitpick](https://github.com/apache/couchdb-config/pull/10/files#r74124401) (feel free to ignore if you disagree) it looks good. +1 from me.

davisp commented:
Good catch on that function name. I'll change that to subscribe.

davisp commented:
@iilyak I changed the function to subscribe and added a test to show that gen_event_EXIT is handled properly. I'll wait for another +1 on that new test before merging.

iilyak commented:
All tests pass locally. +1
```
module 'config'
  config: to_integer_test...ok
  config: to_float_test...ok
  module 'config_tests'
    Config get tests
      config_tests: config_get_test_...ok
      config_tests: config_get_test_...ok
      config_tests: config_get_test_...ok
      config_tests: config_get_test_...ok
      config_tests: config_get_test_...ok
      config_tests: config_get_test_...ok
      config_tests: config_get_test_...ok
      config_tests: config_get_test_...ok
      config_tests: config_get_test_...ok
      [done in 0.557 s]
    Config set tests
      config_tests: config_set_test_...[0.022 s] ok
      config_tests: config_set_test_...[0.001 s] ok
      config_tests: config_set_test_...ok
      [done in 0.109 s]
    Config deletion tests
      config_tests: config_del_test_...ok
      config_tests: config_del_test_...ok
      [done in 0.052 s]
    Configs overide tests
      config_tests:321: should_ensure_in_defaults...ok
      config_tests:329: should_override_options...ok
      config_tests:354: should_create_new_sections_on_override...ok
      config_tests:361: should_win_last_in_chain...ok
      config_tests:336: should_read_default_d...ok
      config_tests:342: should_read_local_d...ok
      config_tests:348: should_read_default_and_local_d...ok
      [done in 0.176 s]
    Config persistent changes
      config_tests:367: should_write_changes...[notice] config: [httpd] port set to 8080 for reason nil
[notice] config: [httpd] bind_address deleted for reason "8080"
[0.024 s] ok
      config_tests:377: should_ensure_default_wasnt_modified...[0.001 s] ok
      config_tests:384: should_ensure_written_to_last_config_in_chain...ok
      [done in 0.095 s]
    Test config with no files
      config_tests: config_no_files_test_...ok
      config_tests: config_no_files_test_...ok
      config_tests: config_no_files_test_...ok
      [done in 0.046 s]
    Test config_listener behaviour
      config_tests:409: should_handle_value_change...ok
      config_tests:416: should_pass_correct_state_to_handle_config_change...ok
      config_tests:425: should_pass_correct_state_to_handle_config_terminate...ok
      config_tests:436: should_pass_subscriber_pid_to_handle_config_terminate...[0.001 s] ok
      config_tests:443: should_not_call_handle_config_after_related_process_death...[0.255 s] ok
      config_tests:456: should_remove_handler_when_requested...[0.001 s] ok
      config_tests:465: should_remove_handler_when_pid_exits...[0.001 s] ok
      config_tests:494: should_stop_monitor_on_error...[0.001 s] ok
      [done in 0.497 s]
    [done in 1.533 s]
  [done in 1.539 s]
=======================================================
  All 37 tests passed.
```
